
<html>
<head>
<meta http-equiv="Content-type" content="text/html; charset=iso-8859-1">
<meta http-equiv="Content-Style-Type" content="text/css">
<style type="text/css">
BODY{background-color: ffffff;
    font-family: monaco, "courier new", monospace;
     color: #000000}
.input, .result, .error{font-weight: bold}
.input {color: #CC6600}
.result{color: #000099}
.error{color: #dd0000}
</style>
</head>
<body>
<pre>
<span class=result><u>Title</u></span>
<br><br>
     <span class=result>mrleaveoneout</span> -- Leave one (genotype) out (at a time) analysis
<br><br>
<a name="syntax"></a><span class=result><u>Syntax</u></span>
<br><br>
        <span class=result>mrleaveoneout</span> <i>varname_gd</i> <i>varname_gp</i> [<i>if</i>] [<i>in</i>]<span class=result>,</span> <span class=result>gyse(</span><i>varname</i><span class=result>)</span> [<i>options</i>]
<br><br>
    <i>options</i>                 Description
    ----------------------------------------------------------------------------------------------
      <span class=result>astext(</span><i>#</i><span class=result>)</span>             Percentage of plot taken up by areas for text - some trial and error
                              may be required
      <span class=result>genotype(</span><i>varname</i><span class=result>)</span>     Variable to label genotypes, usually containing RSIDs
      <span class=result>gxse(</span><i>varname</i><span class=result>)</span>         Variable with genotype-phenotype (SNP-exposure) standard errors (if
                              required by method)
      <span class=result>gyse(</span><i>varname</i><span class=result>)</span>         Variable with genotype-disease (SNP-outcome) standard errors
      <span class=result>metanopts(</span><i>string</i><span class=result>)</span>     Options passed to <span class=result>metan</span> for the plot
      <span class=result><u>m</u></span><span class=result>ethod(</span><i>string</i><span class=result>)</span>        The method fitted for the analysis
      <span class=result>noplot</span>                Suppresses the plot
      <span class=result>noprint</span>               Suppresses the display of each leave one out model
      <span class=result><u>texts</u></span><span class=result>ize(</span><i>#</i><span class=result>)</span>           Scaling factor for text on plot. I have tried to use sensible numbers
                              here but some trial and error may be required
      <span class=result>*</span>                     Other options passed to the analysis method command
<br><br>
<a name="description"></a><span class=result><u>Description</u></span>
<br><br>
    <span class=result>mrleaveoneout</span> performs leave one out analysis, in which each genotype is omitted in turn from
    the analysis.  A plot of the estimates is shown by default.
<br><br>
    For multiple exposure models such as MVMR and MVMR-Egger the estimate is collected for the
    first phenotype.
<br><br>
    <i>varname_gd</i> is a variable containing the genotype-disease (SNP-outcome) association estimates.
<br><br>
    <i>varname_gp</i> is a variable containing the genotype-phenotype (SNP-exposure) association
    estimates.
<br><br>
<a name="options"></a><span class=result><u>Options</u></span>
<br><br>
    <span class=result><u>m</u></span><span class=result>ethod(</span><i>string</i><span class=result>)</span> Specifies the method used for the analysis.  Must be one of ivw, egger,
        mregger, mrivw, median, mrmedian, mrmodal, modal, mode, mvmr, mvivw, mvegger.
<br><br>
<a name="examples"></a><span class=result><u>Examples</u></span>
<br><br>
    Using the data provided by Do et al. (2013).
<br><br>
    Setup
        <span class=result>.</span> use https://raw.github.com/remlapmot/mrrobust/master/dodata, clear
<br><br>
    Select observations (<i>p</i>-value with exposure &lt; 10^-8)
        <span class=result>.</span> gen byte sel2 = (ldlcp2 &lt; 1e-25)
<br><br>
    Perform leave one out analysis using the IVW estimator
        <span class=result>.</span> mrleaveoneout chdbeta ldlcbeta if sel2==1, gyse(chdse) genotype(rsid)
<br><br>
    Leave one out analysis using MVMR (collecting the estimate for LDL-c)
        <span class=result>.</span> mrleaveoneout chdbeta ldlcbeta hdlcbeta tgbeta if sel2==1, method(mvmr) gyse(chdse)
            genotype(rsid)
<br><br>
    If you have data which require more than 2 decimal places, use <span class=result>metan</span>'s undocumented <span class=result>dp</span> option,
    e.g. for 4 decimal places specify <span class=result>dp(4)</span>
        <span class=result>.</span> gen chdbeta2 = chdbeta / 100
        <span class=result>.</span> gen chdse2 = chdse / 100
        <span class=result>.</span> mrleaveoneout chdbeta2 ldlcbeta if sel2==1, gyse(chdse2) genotype(rsid) metanopts(dp(4))
<br><br>
<a name="references"></a><span class=result><u>References</u></span>
<br><br>
<a name="do"></a>    Do et al. Common variants associated with plasma triglycerides and risk for coronary artery
        disease. Nature Genetics, 2013, 45, 1345â€“1352.  DOI
<br><br>
<a name="author"></a><span class=result><u>Author</u></span>
<br><br>
INCLUDE help mrrobust-author
</pre>
</body>
</html>
