
<html>
<head>
<meta http-equiv="Content-type" content="text/html; charset=iso-8859-1">
<meta http-equiv="Content-Style-Type" content="text/css">
<style type="text/css">
BODY{background-color: ffffff;
    font-family: monaco, "courier new", monospace;
     color: #000000}
.input, .result, .error{font-weight: bold}
.input {color: #CC6600}
.result{color: #000099}
.error{color: #dd0000}
</style>
</head>
<body>
<pre>
<span class=result><u>Title</u></span>
<br><br>
     <span class=result>mrforest</span> -- Forest plot for MR-Egger type models
<br><br>
<a name="syntax"></a><span class=result><u>Syntax</u></span>
<br><br>
        <span class=result>mrforest</span> <i>varname_gd</i> <i>varname_gdse</i> <i>varname_gp</i> <i>varname_gpse</i> [<i>varname_cov</i>] [<i>if</i>] [<i>in</i>] [<span class=result>,</span>
              <i>options</i>]
<br><br>
    <i>options</i>                 Description
    ----------------------------------------------------------------------------------------------
      <span class=result>astext(</span><i>#</i><span class=result>)</span>             Percentage of plot taken up by areas for text - some trial and error
                              may be required
      <span class=result>effect(</span><i>string</i><span class=result>)</span>        Label for statistics column; default is Estimate
      <span class=result>ivid(</span><i>varname</i><span class=result>)</span>         Variable to label genotypes, usually containing RSIDs
      <span class=result>ividlabel(</span><i>string</i><span class=result>)</span>     Label for genotypes; default is Genotypes
      <span class=result>ivwlabel(</span><i>string</i><span class=result>)</span>      Label for IVW model; default is IVW
      <span class=result>gsort(</span><i>string</i><span class=result>)</span>         how to sort the estimates, if specified must be one of; ascending,
                              descending, or unsorted; default is ascending
      <span class=result><u>l</u></span><span class=result>evel(</span><i>#</i><span class=result>)</span>              set confidence level; default is <span class=result>level(95)</span>
      <span class=result>mreggerlabel(</span><i>string</i><span class=result>)</span>  Label for MR-Egger model; default is MR-Egger
      <span class=result>mrmedianlabel(</span><i>string</i><span class=result>)</span> Label for median model; default is Median
      <span class=result>mrmodallabel(</span><i>string</i><span class=result>)</span>  Label for modal model; default is Modal
      <span class=result>models(</span><i>#</i><span class=result>)</span>             number of models to show (1 IVW, 2 &amp; MR-Egger, 3 &amp; Median, 4 &amp; Modal);
                              default is 2
      <span class=result>modelslabel(</span><i>string</i><span class=result>)</span>   Label for models; default is Summary
      <span class=result>modelsonly</span>            Only show model estimates on plot
      <span class=result>nonote</span>                Suppress note reporting I^2_GX
      <span class=result>nostats</span>               Suppress the statistics column
      <span class=result><u>texts</u></span><span class=result>ize(</span><i>#</i><span class=result>)</span>           Scaling factor for text on plot. I have tried to use sensible numbers
                              here but some trial and error may be required
      <span class=result>zcis</span>                  Use normal distribution CI limits (for IVW and MR-Egger)
<br><br>
      Options passed to other commands:
      <span class=result>ivwopts(</span><i>string</i><span class=result>)</span>       options for IVW estimate from <span class=result>mregger, ivw fe</span>
      <span class=result>mreggeropts(</span><i>string</i><span class=result>)</span>   options for MR-Egger estimate from mregger
      <span class=result>mrivestopts(</span><i>string</i><span class=result>)</span>   options for mrivests used to generate genotype specific ratio
                              estimates and SEs
      <span class=result>mrmedianopts(</span><i>string</i><span class=result>)</span>  options for median estimate from mrmedian
      <span class=result>mrmodal(</span><i>string</i><span class=result>)</span>       options for modal estimate from mrmodal
      <span class=result>nofe</span>                  Remove <span class=result>fe</span> option (fixed effect SE) from the IVW fit
<br><br>
      <span class=result>*</span>                     Other options passed to the metan
<br><br>
<a name="description"></a><span class=result><u>Description</u></span>
<br><br>
    <span class=result>mrforest</span> plots a forest plot for MR-Egger type models.  It is really a wrapper program for a
    call to metan (Harris et al., 2008), which must be installed.
<br><br>
    If you do not already have <span class=result>metan</span>, it can be installed by running: <span class=result>ssc install metan</span>.
<br><br>
    <i>varname_gd</i> is a variable containing the genotype-disease (SNP-outcome) association estimates.
<br><br>
    <i>varname_gdse</i> is a variable containing the genotype-disease (SNP-outcome) association estimate
    standard errors.
<br><br>
    <i>varname_gp</i> is a variable containing the genotype-phenotype (SNP-exposure) association
    estimates.
<br><br>
    <i>varname_gpse</i> is a variable containing the genotype-phenotype (SNP-exposure) association
    estimate standard errors.
<br><br>
    <i>varname_cov</i> is a variable containing the covariances between the genotype-disease
    (SNP-outcome) and genotype-phenotype (SNP-exposure) associations.
<br><br>
<br><br>
<a name="examples"></a><span class=result><u>Examples</u></span>
<br><br>
    Using the data provided by Do et al. (2013) recreate Bowden et al. (2016) Figure 4, LDL-c "All
    genetic variants" (plot in row 2, column 1).
<br><br>
    Setup
        <span class=result>.</span> use https://raw.github.com/remlapmot/mrrobust/master/dodata, clear
<br><br>
    Select observations (<i>p</i>-value with exposure &lt; 10^-8)
        <span class=result>.</span> gen byte sel1 = (ldlcp2 &lt; 1e-8)
<br><br>
    Forest plot of genotype specific IV estimates and IVW and MR-Egger estimates, labelling the
    genotypes with their RSID.
        <span class=result>.</span> mrforest chdbeta chdse ldlcbeta ldlcse if sel1==1, ivid(rsid)
<br><br>
    Having seen the first plot we can now define nicer <i>x</i>-axis labels.
        <span class=result>.</span> mrforest chdbeta chdse ldlcbeta ldlcse if sel1==1, ivid(rsid)
            xlabel(-5,-4,-3,-2,-1,0,1,2,3,4,5)
<br><br>
    Removing the column of estimates from the plot, and sorting in descending order of the IV
    estimates.
        <span class=result>.</span> mrforest chdbeta chdse ldlcbeta ldlcse if sel1==1, ivid(rsid) nostats gsort(descending)
<br><br>
    Showing all 4 models and modifying some labels.
        <span class=result>.</span> mrforest chdbeta chdse ldlcbeta ldlcse if sel1==1, ivid(rsid) models(4) modelslabel(All
            genotypes)
<br><br>
    If you have data which require more than 2 decimal places, use <span class=result>metan</span>'s undocumented <span class=result>dp</span> option,
    e.g. for 4 decimal places specify <span class=result>dp(4)</span>
        <span class=result>.</span> gen chdbeta2 = chdbeta / 100
        <span class=result>.</span> gen chdse2 = chdse / 100
        <span class=result>.</span> mrforest chdbeta chdse ldlcbeta ldlcse if sel1==1, ivid(rsid) dp(4)
<br><br>
<a name="references"></a><span class=result><u>References</u></span>
<br><br>
<a name="bowden"></a>    Bowden J, Davey Smith G, Haycock PC, Burgess S. Consistent estimation in Mendelian
        randomization with some invalid instruments using a weighted median estimator. Genetic
        Epidemiology, 2016, 40, 4, 304-314.  DOI
<br><br>
<a name="do"></a>    Do et al. Common variants associated with plasma triglycerides and risk for coronary artery
        disease. Nature Genetics, 2013, 45, 1345-1352.  DOI
<br><br>
<a name="harris"></a>    Harris RJ, Bradburn MJ, Deeks JJ, Harbord RM, Altman DG, Sterne JAC. metan: fixed- and
        random-effects meta-analysis.  Stata Journal, 2008, 8, 1, 3-28. Link
<br><br>
<a name="author"></a><span class=result><u>Author</u></span>
<br><br>
INCLUDE help mrrobust-author
</pre>
</body>
</html>
