
<html>
<head>
<meta http-equiv="Content-type" content="text/html; charset=iso-8859-1">
<meta http-equiv="Content-Style-Type" content="text/css">
<style type="text/css">
BODY{background-color: ffffff;
    font-family: monaco, "courier new", monospace;
     color: #000000}
.input, .result, .error{font-weight: bold}
.input {color: #CC6600}
.result{color: #000099}
.error{color: #dd0000}
</style>
</head>
<body>
<pre>
<span class=result><u>Title</u></span>
<br><br>
     <span class=result>mreggerplot</span> -- Scatter plot for MR-Egger type models
<br><br>
<a name="syntax"></a><span class=result><u>Syntax</u></span>
<br><br>
        <span class=result>mreggerplot</span> <i>varname_gd</i> <i>varname_gdse</i> <i>varname_gp</i> <i>varname_gpse</i> [<i>if</i>] [<i>in</i>] [<span class=result>,</span> <i>options</i>]
<br><br>
    <i>options</i>               Description
    ----------------------------------------------------------------------------------------------
      <span class=result>egger</span>               MR-Egger estimator for fitted line
      <span class=result>ellipses</span>            marker confidence intervals as ellipses
      <span class=result>errorbars</span>           marker confidence intervals as capped lines
      <span class=result>fe</span>                  fixed effect standard errors for fitted line
      <span class=result>gpci</span>                CIs around G-P associations
      <span class=result>ivw</span>                 IVW line (default is MR-Egger)
      <span class=result>legend(</span><i>string</i><span class=result>)</span>      legend options
      <span class=result>level(</span><i>#</i><span class=result>)</span>            set confidence level; default is <span class=result>level(95)</span>
      <span class=result>linetop</span>             draw fitted line (and CI) on top of points
      <span class=result>mleglabel(</span><i>string</i><span class=result>)</span>   Label for data points in legend; default is Instruments
      <span class=result>median</span>              median estimator for fitted line
      <span class=result>mlabel(</span><i>string</i><span class=result>)</span>      variable containing marker labels
      <span class=result>nolci</span>               do not plot confidence interval around fitted line
      <span class=result>noline</span>              do not plot fitted line
      <span class=result>nomcis</span>              do not plot confidence intervals around markers
      <span class=result><u>penw</u></span><span class=result>eighted</span>         penalized weighted estimator
      <span class=result>re</span>                  random effect estimator for fitted line
      <span class=result>recons</span>              random intercept for fitted line
      <span class=result>reslope</span>             random slope for fitted line
      <span class=result>reps(</span><i>#</i><span class=result>)</span>             number of bootstrap replications to obtain standard error
      <span class=result>seed(</span><i>#</i><span class=result>)</span>             seed for random number generator for bootstrapping to obtain standard
                            error
      <span class=result><u>w</u></span><span class=result>eighted</span>            weighted median estimator
      <span class=result>wmarkers</span>            weighted markers
      <span class=result>*</span>                   Other options passed to the <span class=result>twoway</span> plot
<br><br>
<a name="description"></a><span class=result><u>Description</u></span>
<br><br>
    <span class=result>mreggerplot</span> plots a scatter plot for MR-Egger type models
<br><br>
    <i>varname_gd</i> is a variable containing the genotype-disease (SNP-outcome) association estimates.
<br><br>
    <i>varname_gdse</i> is a variable containing the genotype-disease (SNP-outcome) association estimate
    standard errors.
<br><br>
    <i>varname_gp</i> is a variable containing the genotype-phenotype (SNP-exposure) association
    estimates.
<br><br>
    <i>varname_gpse</i> is a variable containing the genotype-phenotype (SNP-exposure) association
    estimate standard errors.
<br><br>
<a name="examples"></a><span class=result><u>Examples</u></span>
<br><br>
    Using the data provided by Do et al. (2013) recreate Bowden et al. (2016), Figure 4, LDL-c
    "All genetic variants" (plot in row 2, column 1).
<br><br>
    Setup
        <span class=result>.</span> use https://raw.github.com/remlapmot/mrrobust/master/dodata, clear
<br><br>
    Select observations (<i>p</i>-value with exposure &lt; 10^-8)
        <span class=result>.</span> gen byte sel1 = (ldlcp2 &lt; 1e-8)
<br><br>
    Scatter plot of MR-Egger model
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1
<br><br>
    Scatter plot of MR-Egger model labelling outlying genotypes
        <span class=result>.</span> gen mlabvar = rsid if abs(ldlcbeta) &gt; .3
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, mlab(mlabvar) mlabsize(vsmall)
            mlabp(7) mlabc(gs0)
<br><br>
    Scatter plot of MR-Egger model with genotype-phenotype CIs
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, gpci
<br><br>
    Scatter plot of MR-Egger model specifying own titles and legend label
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, title(Investigating pleiotropy:
            MR-Egger model fit) xtitle(Genotype-LDLC associations) ytitle(Genotype-CHD
            associations) mleglabel(Genotypes)
<br><br>
    Scatter plot of IVW model
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, ivw
<br><br>
    Scatter plot of unweighted median model
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, median
<br><br>
    Scatter plot of MR-Egger model using ellipses around points
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, ellipses
<br><br>
    Scatter plot of MR-Egger model adding an IVW line (with legend entry) for comparison
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1
        <span class=result>.</span> mregger chdbeta ldlcbeta [aw=1/(chdse^2)] if sel1==1, ivw
        <span class=result>.</span> addplot : function _b[ldlcbeta]*x if sel1==1, range(0 0.5) lc(gs0) lp(longdash)
            lw(vthin) legend(order(5 "Instruments" 4 "95% CIs" 3 "MR-Egger" 2 "MR-Egger 95% CI" 6
            "IVW") rows(1) size(vsmall))
<br><br>
    Scatter plot of MR-Egger model adding weighted median and modal lines (with legend entries)
    for comparison
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1
        <span class=result>.</span> mrmedian chdbeta chdse ldlcbeta ldlcse if sel1==1, weighted
        <span class=result>.</span> addplot : function _b[beta]*x if sel1==1, range(0 0.5) lc(gs0) lp(shortdash) lw(vthin)
        <span class=result>.</span> mrmodal chdbeta chdse ldlcbeta ldlcse if sel1==1, phi(.25)
        <span class=result>.</span> addplot : function _b[beta]*x if sel1==1, range(0 0.5) lc(gs0) lp(longdash)
            legend(order(5 "Instruments" 4 "95% CIs" 3 "MR-Egger" 2 "MR-Egger 95% CI" 6 "Weighted
            median" 7 "Modal") rows(1) si(vsmall) symx(*.5))
<br><br>
    Scatter plot without any fitted lines
        <span class=result>.</span> mreggerplot chdbeta chdse ldlcbeta ldlcse if sel1==1, ivw noline nolci gpci
            legend(order(4 "Instruments" 3 "95% CI") size(vsmall) rows(1))
<br><br>
<a name="references"></a><span class=result><u>References</u></span>
<br><br>
<a name="bowden"></a>    Bowden J, Davey Smith G, Haycock PC, Burgess S.  Consistent estimation in Mendelian
        randomization with some invalid instruments using a weighted median estimator. Genetic
        Epidemiology, 2016, 40, 4, 304-314.  DOI
<br><br>
<a name="do"></a>    Do et al.Common variants associated with plasma triglycerides and risk for coronary artery
        disease. Nature Genetics, 2013, 45, 1345â€“1352.  DOI
<br><br>
<a name="author"></a><br><br>
<span class=result><u>Author</u></span>
<br><br>
    Tom Palmer, MRC Integrative Epidemiology Unit and Population Health Sciences, University of
        Bristol, UK. tom.palmer@bristol.ac.uk.
<br><br>
    If you find any bugs or have questions please send me an email or create an issue on the
        GitHub repo: https://github.com/remlapmot/mrrobust/issues
</pre>
</body>
</html>
