<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.24">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>mrrobust</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dc55a5b9e770e841cd82e46aadbfb9b0.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-5b4ad623e5705c0698d39aec6f10cf02.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">mrrobust</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-examples" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Examples</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-examples">    
        <li>
    <a class="dropdown-item" href="./examples/mrrobust-examples/index.html">
 <span class="dropdown-text">Examples from the helpfiles</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./examples/spiller-ije-2018-examples/index.html">
 <span class="dropdown-text">IJE article examples</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./examples/markstat-call-R-example/index.html">
 <span class="dropdown-text">Using TwoSampleMR and mrrobust in a Stata Markdown document</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./examples/rmarkdown-call-stata-example/index.html">
 <span class="dropdown-text">Using TwoSampleMR and mrrobust in an R Markdown or Quarto document</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./examples/save-estimates/index.html">
 <span class="dropdown-text">Collecting and exporting results using <code>r(table)</code></span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./examples/mvcommands-example/index.html">
 <span class="dropdown-text">Fitting MVMR models</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-helpfiles" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Helpfiles</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-helpfiles">    
        <li>
    <a class="dropdown-item" href="./helpfiles/mrrobust-helpfile.html">
 <span class="dropdown-text">mrrobust</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mr-helpfile.html">
 <span class="dropdown-text">mr</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrdeps-helpfile.html">
 <span class="dropdown-text">mrdeps</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mregger-helpfile.html">
 <span class="dropdown-text">mregger</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mreggerplot-helpfile.html">
 <span class="dropdown-text">mreggerplot</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mreggersimex-helpfile.html">
 <span class="dropdown-text">mreggersimex</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmedian-helpfile.html">
 <span class="dropdown-text">mrmedian</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmedianobs-helpfile.html">
 <span class="dropdown-text">mrmedianobs</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmodal-helpfile.html">
 <span class="dropdown-text">mrmodal</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmodalplot-helpfile.html">
 <span class="dropdown-text">mrmodalplot</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrratio-helpfile.html">
 <span class="dropdown-text">mrratio</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrivests-helpfile.html">
 <span class="dropdown-text">mrivests</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrforest-helpfile.html">
 <span class="dropdown-text">mrforest</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrfunnel-helpfile.html">
 <span class="dropdown-text">mrfunnel</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmvivw-helpfile.html">
 <span class="dropdown-text">mrmvivw</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrmvegger-helpfile.html">
 <span class="dropdown-text">mrmvegger</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./helpfiles/mrleaveoneout-helpfile.html">
 <span class="dropdown-text">mrleaveoneout</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/remlapmot/mrrobust"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#mrrobust-stata-package-for-two-sample-mendelian-randomization-analyses" id="toc-mrrobust-stata-package-for-two-sample-mendelian-randomization-analyses" class="nav-link active" data-scroll-target="#mrrobust-stata-package-for-two-sample-mendelian-randomization-analyses">mrrobust: Stata package for two-sample Mendelian randomization analyses</a>
  <ul class="collapse">
  <li><a href="#latest-updates" id="toc-latest-updates" class="nav-link" data-scroll-target="#latest-updates">Latest updates</a></li>
  <li><a href="#short-video-introduction" id="toc-short-video-introduction" class="nav-link" data-scroll-target="#short-video-introduction">Short video introduction</a></li>
  <li><a href="#helpfile-examples" id="toc-helpfile-examples" class="nav-link" data-scroll-target="#helpfile-examples">Helpfile examples</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#installing-and-updating-mrrobust" id="toc-installing-and-updating-mrrobust" class="nav-link" data-scroll-target="#installing-and-updating-mrrobust">Installing and updating mrrobust</a>
  <ul class="collapse">
  <li><a href="#use-net-install" id="toc-use-net-install" class="nav-link" data-scroll-target="#use-net-install">1. Use <code>net install</code></a></li>
  <li><a href="#use-the-github-package" id="toc-use-the-github-package" class="nav-link" data-scroll-target="#use-the-github-package">2. Use the <code>github</code> package</a></li>
  <li><a href="#installation-instructions-for-stata-version-12-and-earlier-versions-and-perhaps-stata-version-13" id="toc-installation-instructions-for-stata-version-12-and-earlier-versions-and-perhaps-stata-version-13" class="nav-link" data-scroll-target="#installation-instructions-for-stata-version-12-and-earlier-versions-and-perhaps-stata-version-13">Installation instructions for Stata version 12 and earlier versions (and perhaps Stata version 13)</a></li>
  </ul></li>
  <li><a href="#unit-tests" id="toc-unit-tests" class="nav-link" data-scroll-target="#unit-tests">Unit tests</a></li>
  <li><a href="#authors" id="toc-authors" class="nav-link" data-scroll-target="#authors">Authors</a></li>
  <li><a href="#how-to-cite-the-mrrobust-package" id="toc-how-to-cite-the-mrrobust-package" class="nav-link" data-scroll-target="#how-to-cite-the-mrrobust-package">How to cite the mrrobust package</a></li>
  <li><a href="#collaboration" id="toc-collaboration" class="nav-link" data-scroll-target="#collaboration">Collaboration</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>




<section id="mrrobust-stata-package-for-two-sample-mendelian-randomization-analyses" class="level1">
<h1>mrrobust: Stata package for two-sample Mendelian randomization analyses</h1>
<ul>
<li><a href="#latest-updates">Latest updates</a></li>
<li><a href="#short-video-introduction">Short video introduction</a></li>
<li><a href="#helpfile-examples">Helpfile examples</a></li>
<li><a href="#overview">Overview</a></li>
<li><a href="#installing-and-updating-mrrobust">Installing and updating mrrobust</a></li>
<li><a href="#unit-tests">Unit tests</a></li>
<li><a href="#authors">Authors</a></li>
<li><a href="#how-to-cite-the-mrrobust-package">How to cite the mrrobust package</a></li>
<li><a href="#references">References</a></li>
<li><a href="#collaboration">Collaboration</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
</ul>
<section id="latest-updates" class="level2">
<h2 class="anchored" data-anchor-id="latest-updates">Latest updates</h2>
<p>To obtain the latest update please see the instructions <a href="#installing-and-updating-mrrobust">below</a>.</p>
<ul>
<li>September 2025:
<ul>
<li>Reran certification scripts on StataNow 19.5</li>
</ul></li>
<li>May 2025:
<ul>
<li>Reran certification scripts (a.k.a. tests) on StataNow 18.5</li>
</ul></li>
<li>July 2024:
<ul>
<li>Updated website to mention <a href="https://api.opengwas.io/">OpenGWAS</a> instead of MR-Base</li>
<li>In Stata 18 on a <code>graph twoway</code> plot the default legend position appears to have changed from the 6 o’clock to the 3 o’clock position. Amendments have been made to <code>mreggerplot</code>, <code>mrfunnel</code>, and <code>mrmodalplot</code> to use the 6 o’clock position as the default again.</li>
</ul></li>
<li>January 2024:
<ul>
<li>Reran certification scripts under Stata 18.0</li>
</ul></li>
<li>August 2023:
<ul>
<li>Added a record of the Stata version in the certification scripts</li>
</ul></li>
<li>April 2023:
<ul>
<li>Improved the alt text descriptions for the images in the <em>README</em> and package website, and also centred the images</li>
<li>Remade the <strong>mrrobust</strong> website using <a href="https://quarto.org/"><strong>Quarto</strong></a></li>
<li>Updated the Sanderson et al., bioRxiv, 2020, <a href="https://doi.org/10.1101/2020.04.02.021980">doi</a> reference to its published version Sanderson et al., Statistics in Medicine, 2021, <a href="https://doi.org/10.1002/sim.9133">doi</a></li>
<li>Reran certification scripts under latest Stata 17.0</li>
</ul></li>
<li>February 2023:
<ul>
<li>Updated R Markdown example to use the CRAN version of the <a href="https://cran.r-project.org/package=Statamarkdown"><strong>Statamarkdown</strong></a> package</li>
</ul></li>
<li>September 2022:
<ul>
<li>Updated manual installation instructions</li>
</ul></li>
<li>February 2022:
<ul>
<li>Ran cscripts under Stata 17.0</li>
<li>Updated website examples to run under Stata 17.0</li>
</ul></li>
<li>September 2021:
<ul>
<li>Changed relevant <code>http:</code> URLs to <code>https:</code></li>
<li>Minor edits to the helpfiles</li>
</ul></li>
<li>June 2021:
<ul>
<li>Published an interactive Code Ocean capsule demonstrating the use of the <strong>mrrobust</strong> package <a href="https://doi.org/10.24433/CO.0587524.v1">here</a></li>
<li>By default <code>mrforest</code> now specifies a fixed effect standard error for its IVW estimate</li>
</ul></li>
<li>April 2021:
<ul>
<li>Added the I-squared statistic and its 95% CI to the <code>mregger ..., heterogi</code> output</li>
</ul></li>
<li>February 2021:
<ul>
<li>Fixes to <code>mrforest</code> and <code>mrleaveoneout</code> related to the recent update to <code>metan</code>. <code>mrforest</code> and <code>mrleaveoneout</code> now use <code>metan9</code> instead of <code>metan</code> because of the changes to <code>metan</code> syntax. No change was necessary in the dependency scripts because <code>metan9</code> is also installed with <code>ssc install metan</code></li>
<li>Checked cscripts pass</li>
<li>Checked examples on website run. And changed the 2 examples which use the <strong>TwoSampleMR</strong> R package to use the new ID code for the exposure data</li>
<li>Updated <code>dependency.do</code> to make it more robust to the more frequent updates to <code>metan</code></li>
<li>Updated <code>mrdeps</code> to make it more robust to the more frequent updates to <code>metan</code></li>
</ul></li>
<li>October 2020:
<ul>
<li><code>dependency.do</code> and <code>mrdeps</code> now install the updated version of the <code>moremata</code> package</li>
<li>Checked the cscripts run under Stata 16.1</li>
<li>Added description of Q-statistic as Cochran’s and Ruecker’s for the IVW and MR-Egger models respectively</li>
<li>In various helpfiles added clarification that genotype-disease stands for SNP-outcome (or indeed instrument-outcome) and that genotype-phenotype stands for SNP-exposure (or indeed instrument-exposure) respectively; i.e.&nbsp;the estimates required for the top and bottom of the IV Wald ratio estimate</li>
</ul></li>
<li>August 2020:
<ul>
<li>Added html versions of the helpfiles to the website. These are available from the <em>Helpfiles</em> website menu bar item</li>
<li>Added extra decimal places examples to helpfiles of <code>mrforest</code> and <code>mrleaveoneout</code></li>
<li><code>mrfunnel</code> now includes a legend on its plot</li>
</ul></li>
<li>July 2020:
<ul>
<li>Added <code>gxse()</code> option to <code>mrmvivw</code> to return instrument strength Q<sub>A</sub> statistic for instrument validity in <code>e(Qa)</code> (<a href="#references">Sanderson et al.&nbsp;2019</a>)</li>
<li>The <code>gxse()</code> option additionally returns the Q<sub>x</sub> and conditional F-statistics for each phenotype for instrument strength in <code>e(Qx)</code> and <code>e(Fx)</code> (<a href="#references">Sanderson et al.&nbsp;2020</a>)</li>
<li>Added <code>tdist</code> option to <code>mrmvivw</code> and <code>mrmvegger</code></li>
<li><code>mrmvivw</code> and <code>mrmvegger</code> now ereturn the RMSE in <code>e(phi)</code></li>
<li><code>mregger, ivw</code> now displays the square root of the residual variance (residual standard error) and ereturns this is <code>e(phi)</code></li>
<li>Checked that examples on website still run</li>
<li>Added <code>mrleaveoneout</code> command to perform leave one out analysis</li>
</ul></li>
<li>June 2020:
<ul>
<li>Simplified the outcome variable name in <code>mregger</code> <code>b</code> and <code>V</code> e-returned matrices. Turn this off with new <code>oldnames</code> option</li>
<li>Added basic multivariable MR-Egger command, <code>mrmvegger</code></li>
<li>Added basic multivariable IVW command, <code>mrmvivw</code> (currently command names <code>mvmr</code> and <code>mvivw</code> also work)</li>
</ul></li>
<li>February 2020:
<ul>
<li>Updated contact details</li>
<li>Minor edits to helpfiles, to show examples setting <code>seed()</code> option where helpful</li>
<li>Fixed <code>mregger</code> bug where <code>r(table)</code> was not returned with the <code>gxse</code> or <code>heterogi</code> options. The output for these options now appears before the coefficient table.</li>
<li>Minor amendments to formatting of <code>mregger</code> <code>gxse</code> output</li>
<li><code>mregger</code> now ereturns <code>e(phi)</code>, the scale parameter, in some cases</li>
</ul></li>
<li>January 2020:
<ul>
<li><code>mregger</code> now additionally returns <code>r(table)</code></li>
<li>Certification scripts: added <code>master.do</code> and renamed and edited a few scripts</li>
<li>Added <code>mr</code> command. Commands may now be run as either <code>mr egger ...</code> or as previously <code>mregger ...</code>.</li>
<li><a href="https://academic.oup.com/ije/pages/the_best_of_ije"><em>Best of IJE 2019</em>!</a></li>
<li><code>mrmedian</code>, <code>mrmedianobs</code>, <code>mreggersimex</code>, <code>mrmodal</code>, and <code>mrratio</code> now additionally return the <code>r(table)</code> matrix (the information from the coefficient table)</li>
<li>Added an example showing how you can save and export your estimates using <code>r(table)</code>, see <a href="https://remlapmot.github.io/mrrobust/examples/save-estimates/index.html">here</a></li>
</ul></li>
<li>December 2019:
<ul>
<li>Added <code>Q_GX</code> to ereturn and display output when <code>gxse()</code> option specified to <code>mregger</code></li>
<li>Changed <code>Q_GX</code> and <code>I^2_GX</code> output to use first order weights in <code>mregger</code> output. This matches the output from the <code>mr_egger()</code> function in the <strong>MendelianRandomization</strong> R package. Use the <code>unwi2gx</code> option to report the unweighted statistics.</li>
</ul></li>
<li>July 2019:
<ul>
<li>Checked that examples on website still run</li>
</ul></li>
<li>December 2018:
<ul>
<li>Improved compatibility with the <a href="https://haghish.github.io/github/"><strong>github</strong></a> Stata package, i.e., <strong>mrrobust</strong> and its dependencies can be installed simply by issuing: <code>gitget mrrobust</code>, if you have the <strong>github</strong> package installed. <a href="#2-use-the-github-package">See below for instructions</a>.</li>
<li><code>mrdeps</code> command added for conveniently installing dependencies</li>
</ul></li>
<li>November 2018:
<ul>
<li>Example showing the use of the <strong>TwoSampleMR</strong> R package and <strong>mrrobust</strong> in the same <a href="https://rmarkdown.rstudio.com/">R Markdown script</a> (<code>.Rmd</code> file) is <a href="https://remlapmot.github.io/mrrobust/examples/rmarkdown-call-stata-example/index.html">here</a></li>
<li>Example showing the use of the <strong>TwoSampleMR</strong> R package and <strong>mrrobust</strong> in the same <a href="https://data.princeton.edu/stata/markdown">Stata Markdown</a> script (<code>.stmd</code> file) is <a href="https://remlapmot.github.io/mrrobust/examples/markstat-call-R-example/index.html">here</a></li>
</ul></li>
<li>September 2018:
<ul>
<li>IJE paper published online <a href="https://doi.org/10.1093/ije/dyy195">here</a></li>
</ul></li>
<li>August 2018:
<ul>
<li>Click <a href="https://remlapmot.github.io/mrrobust/examples/spiller-ije-2018-examples/index.html">here</a> for the example code and output from our IJE article</li>
</ul></li>
<li>May 2018:
<ul>
<li>Click <a href="https://remlapmot.github.io/mrrobust/examples/mrrobust-examples/index.html">here</a> for code and output from the examples in the helpfiles</li>
<li>This page is now rendered on GitHub Pages <a href="https://remlapmot.github.io/mrrobust/">here</a></li>
</ul></li>
<li>April 2018:
<ul>
<li><code>mregger</code> now has option <code>radial</code> which implements the radial formulation of the MR-Egger model, and of the IVW model when used with option <code>ivw</code></li>
</ul></li>
</ul>
</section>
<section id="short-video-introduction" class="level2">
<h2 class="anchored" data-anchor-id="short-video-introduction">Short video introduction</h2>
<p>Click <a href="https://drive.google.com/open?id=0B1owQlNgzNcPY0lMSGk0SnFfQWs">here</a> for a short video demonstrating the use of the package.</p>
<p align="center">
<a href="https://drive.google.com/open?id=0B1owQlNgzNcPY0lMSGk0SnFfQWs"><img src="./img/mrconf2017_video_mrforest_screenshot.png" width="528" height="300" alt="A screenshot of a video demonstrating the use of the mrrobust package."></a>
</p>
</section>
<section id="helpfile-examples" class="level2">
<h2 class="anchored" data-anchor-id="helpfile-examples">Helpfile examples</h2>
<p>Click <a href="https://remlapmot.github.io/mrrobust/examples/mrrobust-examples/index.html">here</a> for some of the code and output from the examples in the helpfiles.</p>
<p>Once the package is installed, there is a summary helpfile which can be viewed in Stata with:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">help</span> mrrobust</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This has links to the helpfile for each command, which has an example near the bottom. In these examples you can click on the code to run it.</p>
</section>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>The <strong>mrrobust</strong> package is a collection of commands for performing two-sample Mendelian randomization analyses using summary data of genotype-phenotype and genotype-outcome associations.</p>
<p>Such data can be obtained from repositories such as MR-Base <a href="https://www.mrbase.org" class="uri">https://www.mrbase.org</a> (<a href="#references">Hemani et al.&nbsp;2016</a>).</p>
<p>The package contains the following commands:</p>
<ul>
<li><code>mrdeps</code> installs dependencies for the package</li>
<li><code>mrratio</code> implements the standard instrumental variable ratio (Wald) estimate with a choice of standard errors/confidence intervals</li>
<li><code>mrivests</code> automates calling <code>mrratio</code> on all the selected genotypes in your dataset</li>
<li><code>mregger</code> implements the IVW and MR-Egger regression approaches introduced in <a href="#references">Bowden et al.&nbsp;(2015)</a></li>
<li><code>mreggersimex</code> implements the simulation extrapolation algorithm for the MR-Egger model</li>
<li><code>mreggerplot</code> implements a scatter plot with fitted line (either from IVW, MR-Egger, or weighted median estimators) and confidence interval</li>
<li><code>mrmedian</code> and <code>mrmedianobs</code> implement the unweighted, weighted, and penalized weighted median IV estimators robust to 50% invalid instruments in <a href="#references">Bowden et al.&nbsp;(2016)</a></li>
<li><code>mrmodal</code> implements the zero modal estimator of <a href="#references">Hartwig et al.&nbsp;(2017)</a></li>
<li><code>mrmodalplot</code> plot of density used in modal estimator</li>
<li><code>mrforest</code> implements a forest plot of genotype specific IV estimates and estimates from models (e.g.&nbsp;IVW and MR-Egger)</li>
<li><code>mrfunnel</code> funnel plot of genotype specific IV estimates</li>
<li><code>mr</code> acts as a primary command, e.g.&nbsp;so the other commands can be run as <code>mr egger ...</code> as well as <code>mregger ...</code></li>
<li><code>mrmvivw</code> (<code>mvmr</code>, <code>mvivw</code>) implements the multivariable IVW model</li>
<li><code>mrmvegger</code> implements the multivariable MR-Egger model</li>
<li><code>mrleaveoneout</code> implements leave one (genotype) out analysis</li>
</ul>
</section>
<section id="installing-and-updating-mrrobust" class="level2">
<h2 class="anchored" data-anchor-id="installing-and-updating-mrrobust">Installing and updating mrrobust</h2>
<p>To install <strong>mrrobust</strong> in Stata versions 13 and later you have two choices.</p>
<section id="use-net-install" class="level3">
<h3 class="anchored" data-anchor-id="use-net-install">1. Use <code>net install</code></h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>net install mrrobust, from(<span class="st">"https://raw.github.com/remlapmot/mrrobust/master/"</span>) <span class="kw">replace</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>mrdeps</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>In this code <code>mrdeps</code> installs the dependencies. These are <code>addplot</code>, <code>kdens</code>, and <code>moremata</code> packages (all by Ben Jann), the <code>heterogi</code> command (Orsini et al.), the <code>metan</code> command (Harris et al.), and the <code>grc1leg</code> command (Wiggins).</p>
<p>If you have previously installed the package and the <code>net install</code> command above fails with an error message that there are two copies of the package installed simply run <code>adoupdate</code>.</p>
<p>To check if there is an update available to any of your user-written Stata packages run <code>adoupdate</code>. To update <strong>mrrobust</strong> run:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adoupdate mrrobust, update</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To uninstall <strong>mrrobust</strong>, issue in Stata:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ado uninstall mrrobust</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If this fails with an error message mentioning that you have <em>“multiple citations/instances of the package installed”</em> simply issue <code>adoupdate mrrobust</code>. This should leave you with the most recent version of the package you previously installed. You can then run <code>ado uninstall mrrobust</code>.</p>
</section>
<section id="use-the-github-package" class="level3">
<h3 class="anchored" data-anchor-id="use-the-github-package">2. Use the <code>github</code> package</h3>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>net install github, from(<span class="st">"https://haghish.github.io/github/"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>gitget mrrobust</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This automatically installs the dependencies.</p>
<p>To update the package issue:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>github update mrrobust</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To uninstall mrrobust issue:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>github uninstall mrrobust</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="installation-instructions-for-stata-version-12-and-earlier-versions-and-perhaps-stata-version-13" class="level3">
<h3 class="anchored" data-anchor-id="installation-instructions-for-stata-version-12-and-earlier-versions-and-perhaps-stata-version-13">Installation instructions for Stata version 12 and earlier versions (and perhaps Stata version 13)</h3>
<p>The <code>net install</code> syntax for installing <code>mrrobust</code> does not work under Stata version 12 and earlier because this webpage has an address starting with https rather than http. In such cases you need to do a manual installation.</p>
<section id="to-download-and-install-mrrobust-manually" class="level4">
<h4 class="anchored" data-anchor-id="to-download-and-install-mrrobust-manually">To download and install mrrobust manually:</h4>
<ul>
<li><p>Click the green “Clone or download” button at the top of the GitHub repository <a href="https://github.com/remlapmot/mrrobust">here</a> and download as a zip archive or click this direct <a href="https://github.com/remlapmot/mrrobust/archive/refs/heads/master.zip">link</a>.</p></li>
<li><p>In your file explorer extract the zip archive and find its filepath, e.g.&nbsp;<code>C:\Users\tom\Downloads\mrrobust-master\mrrobust-master</code></p></li>
<li><p>In Stata run</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>net install mrrobust, from(<span class="st">"C:\Users</span><span class="ch">\t</span><span class="st">om\Downloads\mrrobust-master\mrrobust-master"</span>) <span class="kw">replace</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div></li>
</ul>
<p>The installation commands for the other dependencies should work. However, if you need to install them manually their zip archives are available at the following links (extract the files from the downloaded zip archives and place them in your <code>PERSONAL</code> directory on your <code>adopath</code>):</p>
<ul>
<li><p>the <strong>moremata</strong> package is available as a zip file <a href="http://fmwww.bc.edu/repec/bocode/m/moremata.zip">here</a></p></li>
<li><p>the <code>addplot</code> command is available <a href="http://fmwww.bc.edu/repec/bocode/a/addplot.zip">here</a></p></li>
<li><p>the <code>heterogi</code> command is available <a href="https://ideas.repec.org/c/boc/bocode/s449201.html">here</a></p></li>
<li><p>the <code>kdens</code> command is available <a href="http://fmwww.bc.edu/repec/bocode/k/kdens.zip">here</a></p></li>
<li><p>the <code>metan</code> command is available <a href="https://ideas.repec.org/c/boc/bocode/s456798.html">here</a></p></li>
<li><p>the <code>grc1leg</code> command can be installed with</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode stata code-with-copy"><code class="sourceCode stata"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>net install grc1leg, from(<span class="st">"https://www.stata.com/users/vwiggins"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div></li>
</ul>
</section>
</section>
</section>
<section id="unit-tests" class="level2">
<h2 class="anchored" data-anchor-id="unit-tests">Unit tests</h2>
<p>As far as I know, and unlike R which has the <strong>testthat</strong> package (and other testing packages), there is no recognised standard for writing unit tests for Stata commands. StataCorp. refer to do-files with tests as cscripts (certification scripts). I publish my cscripts (and their log files of output) in the <a href="https://github.com/remlapmot/mrrobust/tree/master/cscripts">cscripts directory</a>.</p>
</section>
<section id="authors" class="level2">
<h2 class="anchored" data-anchor-id="authors">Authors</h2>
<p>Tom Palmer <a href="mailto:tom.palmer@bristol.ac.uk" class="email">tom.palmer@bristol.ac.uk</a>, Wesley Spiller, Neil Davies</p>
</section>
<section id="how-to-cite-the-mrrobust-package" class="level2">
<h2 class="anchored" data-anchor-id="how-to-cite-the-mrrobust-package">How to cite the mrrobust package</h2>
<p>Spiller W, Davies NM, Palmer TM. Software Application Profile: mrrobust — A tool for performing two-sample summary Mendelian randomization analyses. International Journal of Epidemiology, 2019, 48, 3, 684–690. <a href="https://doi.org/10.1093/ije/dyy195" class="uri">https://doi.org/10.1093/ije/dyy195</a>.</p>
<p>Thank you to all our users who have cited <strong>mrrobust</strong>. We made <a href="https://academic.oup.com/ije/pages/the_best_of_ije"><em>The Best of IJE 2019</em></a>!</p>
</section>
<section id="collaboration" class="level2">
<h2 class="anchored" data-anchor-id="collaboration">Collaboration</h2>
<p>If you would like to extend the code or add new commands I am open to receiving pull requests on GitHub or send me an email to <a href="mailto:tom.palmer@bristol.ac.uk" class="email">tom.palmer@bristol.ac.uk</a>.</p>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<div class="hanging-indent">
<p>Bowden J, Davey Smith G, Burgess S. Mendelian randomization with invalid instruments: effect estimation and bias detection through Egger regression. International Journal of Epidemiology, 2015, 44, 2, 512–525. <a href="https://dx.doi.org/10.1093/ije/dyv080" class="uri">https://dx.doi.org/10.1093/ije/dyv080</a>.</p>
<p>Bowden J, Davey Smith G, Haycock PC, Burgess S. Consistent estimation in Mendelian randomization with some invalid instruments using a weighted median estimator. Genetic Epidemiology, 2016, 40, 4, 304–314. <a href="https://dx.doi.org/10.1002/gepi.21965" class="uri">https://dx.doi.org/10.1002/gepi.21965</a>.</p>
<p>Hartwig FP, Davey Smith G, Bowden J. Robust inference in two-sample Mendelian randomisation via the zero modal pleiotropy assumption. International Journal of Epidemiology, 2017, 46, 6, 1985–1998. <a href="https://doi.org/10.1093/ije/dyx102" class="uri">https://doi.org/10.1093/ije/dyx102</a>.</p>
<p>Hemani G et al.&nbsp;The MR-Base platform supports systematic causal inference across the human phenome. eLife, 2018, 7:e34408. <a href="https://doi.org/10.7554/eLife.34408.001" class="uri">https://doi.org/10.7554/eLife.34408.001</a>.</p>
<p>Sanderson E, Davey Smith G, Windmeijer F, Bowden J. An examination of multivariable Mendelian randomization in the single-sample and two-sample summary data settings. International Journal of Epidemiology, 2019, 48, 3, 713–727. <a href="https://doi.org/10.1093/ije/dyy262" class="uri">https://doi.org/10.1093/ije/dyy262</a>.</p>
<p>Sanderson E, Spiller W, Bowden J. Testing and correcting for weak and pleiotropic instruments in two-sample multivariable Mendelian randomization. Statistics in Medicine, 2021, 40, 25, 5434–5452. <a href="https://doi.org/10.1002/sim.9133" class="uri">https://doi.org/10.1002/sim.9133</a>.</p>
</div>
</section>
<section id="acknowledgements" class="level2">
<h2 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h2>
<p>Thanks for helpful feedback and suggestions to (in no particular order): Jasmine Khouja, Michael Holmes, Caroline Dale, Amy Taylor, Rebecca Richmond, Judith Brand, Yanchun Bao, Kawthar Al-Dabhani, Michalis Katsoulis, Ghazaleh Fatemifar, Lai-Te Chen, Sean Harrison, Emma Anderson, Cassianne Robinson-Cohen, Alisa Kjaergaard, and Steve Burgess.</p>
<!-- For generating the html versions of the helpfiles -->


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>